pydicom/dataset.py

```diff
@@ -804,40 +756,58 @@ def convert_pixel_data(self):
-                try:
-                    pixel_array = x.get_pixeldata(self)
-                    self._pixel_array = self._reshape_pixel_array(pixel_array)
-                    if x.needs_to_convert_to_RGB(self):
-                        self._pixel_array = self._convert_YBR_to_RGB(
-                            self._pixel_array
-                        )
-                    successfully_read_pixel_data = True
-                    break
-                except Exception as e:
-                    logger.debug("Trouble with", exc_info=e)
-                    last_exception = e
-                    continue
-            if not successfully_read_pixel_data:
-                handlers_tried = " ".join(
-                    [str(x) for x in pydicom.config.image_handlers])
-                logger.info("%s did not support this transfer syntax",
-                            handlers_tried)
-                self._pixel_array = None
-                self._pixel_id = None
-                if last_exception:
-                    raise last_exception
-                else:
-                    msg = ("No available image handler could "
-                           "decode this transfer syntax {}".format(
-                               self.file_meta.TransferSyntaxUID.name))
-                    raise NotImplementedError(msg)
-            # is this guaranteed to work if memory is re-used??
-            self._pixel_id = id(self.PixelData)
+        if already_have:
+            return
+
+        # Find all pixel data handlers that support the transfer syntax
+        suitable_handlers = [hh for hh in pydicom.config.image_handlers
+                             if hh and hh.supports_transfer_syntax(self)]
+
+        # No suitable handlers are available
+        if not suitable_handlers:
+            raise NotImplementedError(
+                "Unable to decode pixel data with a transfer syntax UID of "
+                "'{0}' ({1}) as there are no suitable pixel data handlers "
+                "available.".format(self.file_meta.TransferSyntaxUID,
+                                    self.file_meta.TransferSyntaxUID.name)
+            )
+
+        last_exception = None
+        for handler in suitable_handlers:
+            try:
+                # Use the handler to get a 1D numpy array of the pixel data
+                arr = handler.get_pixeldata(self)
+                self._pixel_array = reshape_pixel_array(self, arr)
+
+                # Some handler/transfer syntax combinations may need to
+                #   convert the colour space from YCbCr to RGB
+                if handler.needs_to_convert_to_RGB(self):
+                    self._pixel_array = convert_colour_space(self._pixel_array,
+                                                             'YBR_FULL',
+                                                             'RGB')
+
+                # is this guaranteed to work if memory is re-used??
```

Comment:
The id is guaranteed to persist through the object lifetime - what memory re-use is meant here?

Reply:
I have no idea, this is left over from the original code. I'm not an expert on `id` so I left it as is.

Reply:
That was written when I didn't understand id that well either.  I think it is safe to remove.